crmsh/bootstrap.py

```diff
@@ -1692,6 +1693,106 @@ def init():
+    UPDATE_CMD = "echo {} `date +%s` >> {}".format(utils.this_node(), JOIN_LOCK_FILE)
+
+    def __init__(self, target_node):
+        """
+        Init function
+        """
+        self.thread_stop_event = None
+        self.target_node = target_node
+
+    def _run(self, cmd):
+        """
+        Wrapper utils.run_cmd_on_remote, exit when meeting error
+        """
+        rc, out, err = utils.run_cmd_on_remote(cmd, self.target_node)
+        if rc != 0 and err:
+            error(err)
+        return rc, out
+
+    def _detect(self):
+        """
+        Detect whether target node has the lock file
+        """
+        rc, _ = self._run(self.DETECT_CMD)
+        return rc == 0
+
+    def _wait(self):
+        """
+        Wait 15 seconds while detect target node has lock file
+        """
+        warned_once = False
+        origin_timestamp = None
+
+        while self._detect():
+            # tail the last line in lock file
+            rc, out = self._run(self.TAIL_CMD)
+            if rc == 0 and out:
+                try:
+                    current_node, timestamp = out.split()
+                except ValueError:
+                    error("Invalid format for {}(should be \"node timestamp\")".format(self.JOIN_LOCK_FILE))
+
+                if origin_timestamp:
+                    res = int(timestamp) - int(origin_timestamp)
+                    # current join node might has exceptions
+                    if res < 10:
+                        # remove lock file on target node
+                        self._run(self.RM_CMD)
+                        break
+                else:
+                    origin_timestamp = timestamp
+
+                if not warned_once:
+                    warned_once = True
+                    warn("Node {} still joining, wait...".format(current_node))
+            time.sleep(15)
+
+    def _update(self):
+        """
+        Update timestamp(10s interval) in lock file on target node, format is:
+            nodename timestamp
+        """
+        while not self.thread_stop_event.is_set():
+            # TODO utils.run_cmd_on_remote have issue to run in thread
+            if not invoke("ssh root@{} \"{}\"".format(self.target_node, self.UPDATE_CMD)):
+                error("Can't invoke {} on {}".format(self.UPDATE_CMD, self.target_node))
+            time.sleep(10)
+
+    def lock(self):
+        """
+        Create lock file on target node
+        """
+        # wait if other node was joining
+        self._wait()
+
+        # touch lock file
+        self._run(self.TOUCH_CMD)
+
+        self.thread_stop_event = threading.Event()
+        # start thread to update timestamp in lock file
+        th = threading.Thread(target=self._update)
+        th.start()
+
+    def unlock(self):
+        """
+        Stop the update thread and remove the lock file on target node
+        """
+        self.thread_stop_event.set()
+        self._run(self.RM_CMD)
```

Comment:
We should run the command before setting the event flag, right?

Reply:
I think it's not a big issue, besides that remove first might lead to update timstamp command failed